// @tx/playwright-core/dist/core/actions/navigation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.navigateTo = navigateTo;
exports.refreshPage = refreshPage;
exports.navigatBack = navigatBack;
exports.navigatForward = navigatForward;
const BrowserManager_1 = require("../BrowserManager");
async function navigateTo(url) {
    const page = await BrowserManager_1.BrowserManager.getPage();
    await page.goto(url);
}
async function refreshPage(options) {
    const page = await BrowserManager_1.BrowserManager.getPage();
    await page.reload({
        waitUntil: options?.waitUntil || 'load',
        timeout: options?.timeout,
    });
}
async function navigatBack(options) {
    const page = await BrowserManager_1.BrowserManager.getPage();
    await page.goBack({
        waitUntil: options?.waitUntil || 'load',
        timeout: options?.timeout,
    });
}
async function navigatForward(options) {
    const page = await BrowserManager_1.BrowserManager.getPage();
    await page.goForward({
        waitUntil: options?.waitUntil || 'load',
        timeout: options?.timeout,
    });
}
