// playwright-selfheal-ai/dist/src/NextGenHeal(AI)/SelfHealingPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SelfHealingPage = void 0;
const LocatorStore_1 = require("./LocatorStore");
const SelfHealingHelper_1 = require("./SelfHealingHelper");
const Logger_1 = require("./Logger");
class SelfHealingPage {
    constructor(page, config = {}) {
        this.page = page;
        this.config = config;
        Logger_1.Logger.configure(config.logLevel || "info");
        this.store = new LocatorStore_1.LocatorStore();
        this.helper = new SelfHealingHelper_1.SelfHealingHelper(this.page, this.store);
    }
    /**
     * Returns a proxy Locator wrapper.
     * Any Playwright Locator action (click, fill, type, press, hover, check, selectOption, etc.)
     * will automatically run through self-healing logic.
     */
    locator(selector) {
        const self = this;
        return new Proxy({}, {
            get(_target, propKey) {
                return async (...args) => {
                    // Resolve the element with healing
                    const resolved = await self.helper.find(selector);
                    // Get the method being called (e.g., click, fill, type, hover, etc.)
                    const fn = resolved[propKey];
                    if (typeof fn !== "function") {
                        throw new Error(`[SelfHealing] Locator has no method: ${propKey}`);
                    }
                    Logger_1.Logger.info(`[SelfHealing] Executing action: ${propKey} on selector: ${selector}`);
                    return fn.apply(resolved, args);
                };
            },
        });
    }
}
exports.SelfHealingPage = SelfHealingPage;
