// @tx/playwright-core/src/core/actions/navigation.ts
import { BrowserManager } from "../BrowserManager";

interface RefreshOptions {
    waitUntil?: 'load' | 'domcontentloaded' | 'networkidle';
    timeout?: number;
    ignoreCache?: boolean;
}
export async function navigateTo(url: string) {
    const page = await BrowserManager.getPage();
    await page.goto(url);
}

export async function refreshPage(options?: RefreshOptions) {
    const page = await BrowserManager.getPage();
    await page.reload({
        waitUntil: options?.waitUntil || 'load',
        timeout: options?.timeout,
    });
}

export async function navigatBack(options?: RefreshOptions) {
    const page = await BrowserManager.getPage();
    await page.goBack({
        waitUntil: options?.waitUntil || 'load',
        timeout: options?.timeout,
    });
}

export async function navigatForward(options?: RefreshOptions) {
    const page = await BrowserManager.getPage();
    await page.goForward({
        waitUntil: options?.waitUntil || 'load',
        timeout: options?.timeout,
    });
}
